feat: ✨ pull requests chapter (#102)
seedcase-project/guidebook: workflows/prs.qmd

# Pull requests

Pull requests (PRs) are how we propose, review, and discuss changes
before they become part of a project on GitHub. In fact, we **require**
that all changes to a repository on GitHub go through a pull request. We
use pull requests for several reasons. They:

> 1. Help to maintain the code and text quality and minimise the risk of

Author: ```suggestion
1. Help to maintain code and text quality and minimise the risk of
```

   project.

As the American software developer (and founder of Stack Exchange) Jeff
Atwood [wrote](https://blog.codinghorror.com/code-reviews-just-do-it/):

> "**Peer code reviews are the single biggest thing you can do to improve your
code**. If you're not doing code reviews right now with another
developer, you're missing a lot of bugs in your code and cheating
yourself out of some key professional development opportunities. As far
as I'm concerned,
**my code isn't done until I've gone over it with a fellow developer**."

This section covers the practices we follow for creating and reviewing
pull requests.

## What is a pull request?

To understand pull requests, it helps to first understand *branches*.
When you work on a project in Git, there is a default branch (usually
called `main`) that holds the current, stable version of the project.
Instead of making changes directly to `main`, you create a *branch* ---
think of it like making a copy of a document so you can draft changes
without touching the original. You work on your branch independently,
and when you are happy with your changes, you propose merging them back

> into `main`. That proposal is called a **pull request** (PR).

Author: ```suggestion
without touching the original. A branch allows you to work on your changes independently,
and, when you are happy with your changes, you propose merging (or adding) them
into the project's `main` branch. That proposal is called a **pull request** (PR).
```

A PR gives your collaborators the opportunity to review your changes,
ask questions, and suggest improvements before the changes become part
of the project. This review process helps catch mistakes, maintain
quality, and keep your collaborators informed about what is changing and
why.

The diagram below shows this in practice. A person creates a branch
called `docs/add-pr-guide` to work on their changes. The dots represent
[commits](commits.qmd), which are saved snapshots of their work. When

> ready, they open a PR. After review and approval, their changes are

Author: ```suggestion
[commits](commits.qmd), which are saved snapshots of their work. When they have completed their
changes, they open a PR. After review and approval, the changes are
```

merged back into the `main` branch of the project.

```{mermaid}
---
config:
  theme: 'forest'
  gitGraph:
    showCommitLabel: false
---
gitGraph
    commit
    commit
    branch docs/add-pr-guide
    checkout docs/add-pr-guide
    commit
    commit
    commit tag:"Make PR"
    checkout main
    merge docs/add-pr-guide
    commit
```

Generally, the PR process follows these steps:

- Create a branch for your changes.
- Make your changes and commit them to the branch.
- Open a pull request on GitHub to propose merging your changes into
  `main`.
- Review and discuss the changes with your collaborators.
- If the changes are approved, merge the PR into `main`. If not, make
  further changes, update the PR, and request another review.

This workflow is called the
[GitHub Flow](https://decisions.seedcase-project.org/why-github-flow/)

> and it is a widely used and effective collaboration strategy for

Author: ```suggestion
and is a widely used and effective collaboration strategy for
```

The syntax looks like this:

```` text
```suggestion
the corrected or improved text here
```
````

When leaving comments, be specific about what you think should change
and why. Avoid vague feedback like "this doesn't look right" and instead
explain what the issue is and how it could be improved.

### Comment, request changes, or approve

When you finish your review, GitHub gives you three options for
submitting it:

- **Comment**: General feedback or questions that don't block merging.
  Good for minor suggestions the author can choose to address.
- **Request changes**: Issues that must be resolved before merging. This
  signals that further work is needed and prevents accidental merging.
- **Approve**: The changes are approved and ready to be merged.

If a reviewer requests changes, then you will need to address the
feedback, update the PR, and re-request a review by clicking the
"Re-request review" button in the right sidebar of the PR page. This
back and forth continues until the PR is approved and ready to merge.

### Dealing with merge conflicts

Sometimes, someone else may have made changes to the same files or lines

> as the changes you've made. This will lead to *merge conflicts* as Git

Author: ```suggestion
as the changes you've made and merged them to `main`. This will lead to *merge conflicts* as Git
```

- If the changes are approved, merge the PR into `main`. If not, make
  further changes, update the PR, and request another review.

This workflow is called the
[GitHub Flow](https://decisions.seedcase-project.org/why-github-flow/)
and it is a widely used and effective collaboration strategy for
projects of all sizes.

## Creating a pull request

### How to create a PR

There are a few ways to make a PR. We use two ways: through VS Code and
through the GitHub website. Using the GitHub website is a fairly
straightforward way of making a PR, as you can do it through the pull
request tab of the repository by clicking the "New pull request" button
and following the instructions there.

> Doing it through VS Code, however, is a faster way, though it requires

Author: ```suggestion
Creating a PR through VS Code, however, is usually faster, though it requires
```

"Re-request review" button in the right sidebar of the PR page. This
continues until the PR is approved and ready to merge.

### Dealing with merge conflicts

Sometimes, someone else may have made changes to the same files or lines
as the changes you've made. This will lead to *merge conflicts* as Git
can't determine what to do with the conflicting changes. Dealing with
merge conflicts requires manual editing and resolving to continue.

There are two main ways to resolve merge conflicts:

- Use the GitHub interface to resolve the conflicts directly on the
  website. This is the easiest way for simple conflicts. However, GitHub
  can't handle more complicated conflicts, so you need to resolve it on
  your computer and push the resolved changes to GitHub.
- Pull any changes from the remote repository to your local repository
  by using either `git` (e.g. `git pull origin main`) or through VS
  Code's Git features. You can use the command palette and search for
  `Git: Pull from` and then pull from `origin/main`. Then you can
  resolve the conflicts with the help of VS Code and with manual
  editing. After you've resolved the conflicts, you can push the changes

>   to GitHub to update the PR.

Author: ```suggestion
- Use the GitHub interface to resolve the conflicts directly on the
  website. This is the easiest way for simple conflicts. 
- Resolve the merge conflicts locally. GitHub
  can't handle more complicated conflicts, so you sometimes need to resolve it on
  your computer and push the resolved changes to GitHub. 

To handle a merge conflict locally, pull any changes from the remote repository to your local repository by using either `git` (e.g. `git pull origin main`) or through VS
  Code's Git features. You can use the command palette and search for
  `Git: Pull from` and then pull from `origin/main`. Then, you can
  resolve the conflicts with the help of VS Code and with manual
  editing. After you've resolved the conflicts, you can push the changes
  to GitHub to update the PR.
```

Author: I get that; yeah, the "new review experience" was acting up yesterday.
